### ISSUE TITLE
Queuer / Worker versioning incompatibilties

### ISSUE THREAD

**Issue by in03** [in03/Resolve-Proxy-Encoder#22]
## Issue
Currently any workers subscribed to the broker will take supported jobs. If there are any changes to the project structure or task implementations, the worker will take the job and do what it thinks it should do. This can lead to `unregistered task` errors, incomplete jobs or even jobs that are completed but done wrong without any indication.

### Easy
- Automatically update to the newest release on both queuer and worker startup (if auto-update is set in settings)

- Warn user (if auto-update is not set) when new version is available (like pip does for Python) and briefly explain potential incompatibility issues

### Hard

- Warn the user before queuing If multiple versions of workers are detected as connected to the broker

- Implement a versioning system that prevents workers from taking jobs from newer versions of the software. Log to the console when a job has been ignored and warn the user they should update to the latest version.

**Comment by in03** [comment]
Maybe look into this to print some basic git versioning in the CLI? Expand on that a little more?
https://github.com/spapas/python-git-info

**Comment by in03** [comment]
This is looking pretty doable. Each package installed from github with pip has a file called `direct_url.json` in the site-package's `dist-info` folder with this info:
```
{"url": "https://github.com/user/myrepo", "vcs_info": {"commit_id": "dbd1927adc3f6d830f12fa10656f1557e86024e7", "vcs": "git"}}
```

This info is programmatically accessible using `pkg_resources.get_distribution(package_name).get_metadata("direct_url.json")`
This is all the info we need to compare the installed version's commit ID to git history and determine if there are any outdated / mismatched queuers or workers.

As far as I know, this will only work for git installs. We'll need another method for package releases (once we get that ball rolling properly), but since we'll be using semantic versioning (hopefully automatically), we can just compare package versions. 

Perhaps we can run tests for breaking changes and use each major release to separate incompatible versions. Then we disallow different major versions in-app and prompt for update.

**Comment by in03** [comment]
Unfortunately, we can't get commit history directly from git without cloning the repository. 
We can try cloning only the last few commits to save time and space, but if this should run every startup, we want this to be as quick as possible. We will also need to parse this data in a way that will be easy for us to use.

Otherwise, we know we're using GitHub and we can get results using their API. If we ever change from GitHub, we'll need to rewrite all of this version control stuff, but hopefully if that ever happens, we'll be running proper semantic versioning and be checking that way instead of with git commits. If we use GraphQL we get a speed advantage over REST. Since we're not actually cloning any of the source files, we can potentially pull a lot more commit history than the cloning method.

What probably makes the most sense is having one method to compare for incompatible versions and another method to check for updates. On startup, we use the GraphQL API to grab the last 10 or so commits. Then we can get a nice number of commits we're behind on until 10:
 
```Update available. You are 3 commits behind. ```  
```Update available. You are at least 10 commits behind.```

Then for checking incompatible versions, we simply check the `direct_url.json` and make sure the commits are the same. If they're not the same, or there are any errors on the way, we can chalk it up to incompatibility and warn the user with a more serious prompt. We'll have to find a way to run the worker commit check regularly. Maybe before every group of tasks? Just in case we have workers running for a long period of time between shut downs, and therefore startup checks? We'll need to include the case of having multiple possible versions of queuers, some potentially older than the workers.


## Git clone last three commits
`git clone -b [branch name] --single-branch [repo url] --depth=3`

## GitHub GraphQL API
```
{
  repository(name: "resolve-proxy-encoder", owner: "in03") {
    defaultBranchRef {
      target {
        ... on Commit {
          history(first: 3) {
            nodes {
              message
              committedDate
              oid
              author {
                email
                name
              }
            }
          }
        }
      }
    }
  }
}
```

**Comment by in03** [comment]
# In Smaller Pieces I Reckon...
It's been a while but I've had some thoughts coming back to this.

Prompting to install updates (and the various levels of automation around it), should be considered one distinct mechanism. All we're doing is letting the user know an update is available and potentially installing it. Preventing mismatched queuers / workers from transmitting tasks should be something completely different. 

## Prompting Updates
The earlier ideas on fetching updates should stand pretty well. It's a little clunky asking GitHub, but once we get proper semantic versioning going we may well be able to deprecate this. I'm not too keen on jumping straight into automatic versioning, because I'm making breaking changes all the time at this early stage and I don't want to have heaps of version numbers flying everywhere. Once I get the majority of the bugfixes and enhancements I want done and things calm done, we can look at that.

## Warning and Blocking Version-Mismatches
The second part of this problem is making sure the application can still function properly even if the user can't update old versions so machines are all running the same. I think this needs to come in two parts: warning and blocking. The reason they should be separate solutions is in the timing, _pre-runtime_ and _runtime_. Trying to come up with a catch-all solution that can both block incompatible workers and incorporate clear logging would require a lot of wheel-reinventing. 

Separating the issues into two solutions, **warning** and **blocking** is quite accessible:

### Warning pre-runtime
1. Query Celery for connected workers
2. Run a task on the worker to get local GIT version
3. Return the version to the queuer
4. Tidily print matched-version workers **only** as available workers
5. Print an error to console with any mismatched-version workers as "Incompatible"

### Blocking at runtime
1. Get the git commit SHA
2. In Celery's task configuration, create explicit queues using the git SHA as a suffix.
3. Workers only take tasks from queues matching their local git SHA.

**Comment by in03** [comment]
This is now implemented! 👍  f635cee2978de73546d63460bbef1aebec472339
